src/execution/compiler/operator/hash_join_translator.cpp

```diff
@@ -252,6 +305,60 @@ void HashJoinTranslator::CheckRightMark(WorkContext *ctx, FunctionBuilder *funct
   check_condition.EndIf();
 }
 
+void HashJoinTranslator::CollectUnmatchedLeftRows(FunctionBuilder *function) const {
+  auto *codegen = GetCodeGen();
+
+  // var joinHTIterBase: JoinHashTableIterator
+  auto jht_iter_base = codegen->MakeFreshIdentifier("joinHTIterBase");
+  auto jht_iter_type = codegen->BuiltinType(ast::BuiltinType::JoinHashTableIterator);
+  function->Append(codegen->DeclareVarNoInit(jht_iter_base, jht_iter_type));
+
+  // var joinHTIter = &joinHTIterBase
+  auto jht_iter = codegen->MakeFreshIdentifier("joinHTIter");
+  auto jht_iter_init = codegen->AddressOf(codegen->MakeExpr(jht_iter_base));
+  function->Append(codegen->DeclareVarWithInit(jht_iter, jht_iter_init));
+
+  // while (hasNext()):
+  auto jht = global_join_ht_.GetPtr(codegen);
+  auto jht_iter_expr = codegen->MakeExpr(jht_iter);
+  Loop loop(function, codegen->MakeStmt(codegen->JoinHTIteratorInit(jht_iter_expr, jht)),
+            codegen->JoinHTIteratorHasNext(jht_iter_expr),
+            codegen->MakeStmt(codegen->JoinHTIteratorNext(jht_iter_expr)));
+  {
+    // var buildRow = @joinHTIterGetRow()
+    function->Append(
+        codegen->DeclareVarWithInit(build_row_var_, codegen->JoinHTIteratorGetRow(jht_iter_expr, build_row_type_)));
+
+    auto left_mark = codegen->AccessStructMember(codegen->MakeExpr(build_row_var_), build_mark_);
+
+    // If mark is true, then row was not matched
+    If check_condition(function, left_mark);
+    {
+      // var probeRow : ProbeRow
```

Comment:
These are for build rows, whose probe-side attributes are null?

Reply:
I'm not sure how to parse the question exactly. 

These probeRows are for the buildRows that weren't matched, so for left outer join need to output the buildRow + NULLs. Since, I'm passing a probe row to the `joinConsumer` function which it uses to read the right-side values, I set the probe row values to NULL before passing it.